### DISCUSSION TITLE
Display GPS quality / satellites

### DISCUSSION THREAD

**Discussion by kds112** [devemux86/cruiser#287] [Q&A] [answered]
When I am hiking sometimes the GPS of my smartphone fails and no satellite is locked (a reboot helps in that situation). But I do not recognize it, the arrow just remains at the last known position. Is it possible to implement an option to show (maybe near to the map scale) a satellite symbol, the number of locked satellites, the number of satellites available and the possible error distance in meters? In the app "GPS Status"  all these parameters are available. It would be a great help to see directly on the Cruiser map whether GPS is okay or not.
Thanks a lot!

**Comment by devemux86** [comment]
@kds112 thanks for the suggestion!

It would be indeed a useful feature for everyone, I will put it in my implementation list.
Probably as a new panel at top right.

**Reply by kds112** [reply]
That would be great! Please make it visible not only in Navigation/Follow mode but also in the "normal" map view. Thanks!!

**Reply by kds112** [reply]
By the way, is it possible to show the other top-right panels like height, time, battery in the "normal" map view as well? Not only in Navigation/Follow mode?

**Reply by devemux86** [reply]
@kds112 non-navigation related panels could probably be displayed in normal map view as well.

I could disconnect these panels from the active mode, so they can always appear (if enabled).

**Reply by kds112** [reply]
Wonderful! :-)

**Reply by Ciutnik** [reply]
I often see on maps a "circle of accuracy" drawn around the current position. A larger circle meant less accuracy. Perhaps such a solution would suffice? 

At the moment in Cruiser this circle is fixed, it's part of the arrow (location), and to be honest I'm not sure if it's needed in this form at all.

**Reply by devemux86** [reply]
Current transparent circle is more for better indication of location marker or arrow on the map.

Cruiser already supports an accuracy circle for GPS. It is the same circle with a variable size.
But in the past users did not like it, especially with less good reception when it fills the screen!

**Reply by devemux86** [reply]
By the way we can make current location circle optional, if you want?

**Reply by Ciutnik** [reply]
Oh, I didn't remember that the users wish was a fixed circle size. Of course, I respect that and in fact, with a big position error, it might look bad.

I've gotten used to it by now and would rather not turn it off now 🙂

**Reply by devemux86** [reply]
I remember when Android started, all map apps had variable location accuracy circle.

Is it still popular these days or do you see it less often in map apps?

**Reply by Ciutnik** [reply]
> I remember when Android started, all map apps had variable location accuracy circle.

That's how I remember it, which is why it came to my mind now as a clever solution I've always used 🙂 Google Map still uses it! Also OsmAnd and Locus.

Another thing is that huge accuracy errors happened much more often than now.

**Reply by kds112** [reply]
What about the accuracy circle as an option?

**Reply by devemux86** [reply]
> What about the accuracy circle as an option?

This is easy, as the implementation already exists. :slightly_smiling_face: 

We could have it with 3 options to cover all cases?
- **Default** (fixed circle)
- **Accuracy** (variable circle)
- **Off** (without circle)

**Comment by devemux86** [comment]
We currently have 5 info panels that can work outside navigation:
- Speed
- Average speed
- Altitude
- Current time
- Battery level

I understand that "Current time" and "Battery level" can be used at any time.

But "Speed" / "Average speed" and "Altitude" take their values from the GPS.
So does it make sense to show these 3 when we don't have follow enabled?

**Reply by WolfgangZ42** [reply]
I'd need none of them - but I'm a minimalist ... 🥴

**Reply by ramondevic** [reply]
Perhaps  "Speed" and "Average speed" is not necessary when we don't have follow enabled but "Altitude" could be used at any time like "Current time" and "Battery level"

**Reply by kds112** [reply]
Speed only makes sense in Follow mode, but height is useful all the time. And if GPS does not deliver data you can show n/a.

**Reply by devemux86** [reply]
"Altitude" is currently read from the GPS sensor (normal or mean sea level).

BRouter files contain elevation data, we could ask **offline** BRouter for altitude.
However they probably store elevation info only for the routable network?

cc @afischerdev

**Reply by kds112** [reply]
I would prefer altitude data from the gps sensor...

**Reply by ramondevic** [reply]
Yes me too.

Also, if we don't use BRouter, we cannot see Altitude ?

**Reply by devemux86** [reply]
> if we don't use BRouter, we cannot see Altitude ?

BRouter routing and BRouter elevation are separate services.
If BRouter data are installed, we can query them, regardless of routing.

**Reply by kds112** [reply]
And please add a new optional panel with locked and visible satellites plus accuracy in meters. Thanks!  :-)

**Reply by devemux86** [reply]
You can test the display of altitude / battery level / current time outside navigation and follow in:

- https://github.com/devemux86/cruiser/discussions/299

**Reply by devemux86** [reply]
For those who only want them in navigation / follow, we should probably change the toggle to options like in other settings:
- Always
- Navigation / Follow
- Off

**Comment by devemux86** [comment]
New "Settings | Location | Appearance | Location circle":
- Normal
- Accuracy
- Off

<img src="https://github.com/devemux86/cruiser/assets/3484020/2ba10a47-9223-4976-b620-e1366a614a36" width="33%"/>
<img src="https://github.com/devemux86/cruiser/assets/3484020/b9058a0f-e758-4075-b8c8-3c5710449db7" width="33%"/>
<img src="https://github.com/devemux86/cruiser/assets/3484020/6586d0ae-904e-4069-8e2b-f1ba3823f84d" width="33%"/>

**Comment by devemux86** [comment]
- **Altitude** / **Current time** / **Battery level** can be adjusted to display **Always** or **Navigation / Follow** or **Off**

- **Location circle** was added in location appearance settings with options for **Normal** or **Accuracy** or **Off**

Available for testing in:
- https://github.com/devemux86/cruiser/discussions/299

**Reply by Ciutnik** [reply]
Is it just me, after turning off the accuracy circle, the location arrow also disappeared?

**Reply by devemux86** [reply]
@Ciutnik can you write more details?

In what state: map view or follow mode or navigation?
What option was selected before and what was selected after?

**Reply by Ciutnik** [reply]
Both in navigation, as well as in following and browsing the map. The other circle options don't cause a problem. Interestingly, it happens the same on a Xiaomi phone and a Samsung tablet!

EDIT: When I switch "Location Appearance" to "Normal", the cursor appears.

**Reply by devemux86** [reply]
This is interesting, I cannot reproduce it on native Android and OnePlus.

Can other users test the report?

**Reply by devemux86** [reply]
> EDIT: When I switch "Location Appearance" to "Normal", the cursor appears.

Thanks, with "Advanced" appearance I can reproduce it!
I will see what happens. :slightly_smiling_face:

**Reply by Ciutnik** [reply]
Please remind me the differences between "Location Appearance" | "Normal" vs "Advanced". I must have known it once but... I forgot.

**Reply by devemux86** [reply]
"Advanced" was the initial implementation of location marker + arrow.
The location is rendered in pure OpenGL.

Some devices had rendering issues and although we fixed them,
for compatibility we also offer the "Normal" appearance using icons.

**Reply by Ciutnik** [reply]
Thanks for reminding me.  No wonder I forgot because I've never had problems with an arrow 😉 I always had "Advanced" turned on.

**Reply by devemux86** [reply]
I will see if it is possible to remove the circle in OpenGL option.

Although we should probably only offer the "Normal" location appearance.

**Reply by Ciutnik** [reply]
> Although we should probably only offer the "Normal" location appearance.

Sure. If "Advanced" _(except for a slightly nicer shape for me)_ does not bring much from the user's point of view, it is worth limiting options that are easy to forget what they are for 🙂

**Reply by devemux86** [reply]
OpenGL is complicated, any update requires (not simple) changes in the GL shaders.

With normal appearance, if we want a different marker / arrow, we only have to change the icon.

**Reply by Ciutnik** [reply]
> OpenGL is complicated

It is all the more worth limiting the options so that it is easier for you to embrace the increasingly complex matter that is Cruiser 🙂

**Reply by airald** [reply]
I see always the arrow...

**Reply by devemux86** [reply]
@airald check with location circle "Off" and location appearance "Advanced".

It is probably another option we can remove for less user confusion,
as "Normal" allows greater flexibility and is more simple to maintain.

**Comment by airald** [comment]
Using normal and circle off, the arrow disappeared.

I use advanced, but I tested Normal, I think is enough. Even the position dot is better, its orientation arrow is more visible than in Advanced mode because overlaps with circle.

**Reply by devemux86** [reply]
> Using normal and circle off, the arrow disappeared.

@airald are you sure about the "Normal"? Or it was "Advanced"?

**Reply by airald** [reply]
sorry, I would say Advanced!!!

**Comment by devemux86** [comment]
GPS info will be available in the next Cruiser:

<img src="https://github.com/devemux86/cruiser/assets/3484020/de1c239c-9b66-44a2-bd49-89018b513ec5" width="50%"/><img src="https://github.com/devemux86/cruiser/assets/3484020/76f19aa2-9b42-4bda-88d0-6ab746bc1fb9" width="50%"/>

**Reply by kds112** [reply]
Perfect!!

**Reply by devemux86** [reply]
_Note: use it if you need it, as additional GPS sensor processing may consume more battery._

**Reply by Ciutnik** [reply]
It's good to have. While I understand the need, it's probably not for me 🙂

_By the way, "GPS info" is also quite correct in Polish_ 👍

**Reply by devemux86** [reply]
It is one of those features that can be implemented quite easily. :slightly_smiling_face: 

> By the way, "GPS info" is also quite correct in Polish 👍

@Ciutnik it was translated as "Informacje GPS" or can be "GPS info"?

**Reply by devemux86** [reply]
Info: it works from **Android 7** (Nougat) and later.

**Reply by Ciutnik** [reply]
> It was translated as "Informacje GPS" or can be "GPS info"?

Both are correct and understandable 🙂

Maybe "Informacje GPS" is more officially "dictionary". Since most of the translations are just that, this may be the case this time as well 👍

Oh, in that case "Off" = "Wyłączone"

**Reply by devemux86** [reply]
@Ciutnik all settings with such options use the same shared strings:
- Zawsze
- Nawigacja / Podążanie
- Nigdy

Do they need any improvement?

**Reply by devemux86** [reply]
"GPS info" panel is available for testing in:

- https://github.com/devemux86/cruiser/discussions/299

Please try it on all your devices.
Also test with both **Android** or **Google** location services (Cruiser location settings).

**Reply by Ciutnik** [reply]
> Do they need any improvement?

No no no! I was thinking of a different set of selectable options, for a different case. The one you quoted here is perfect for this use. Sorry, I lost my orientation for a moment 🤪

**Reply by Ciutnik** [reply]
> Please try it on all your devices.

I checked on three devices, each with both location services. Everything is fine as expected 👍

**Reply by kds112** [reply]
In the latest Update 12 the badges work correctly. Any chance to change the order?
I have a problem with the accuracy circle. The accuracy is about 7 meters in the moment (seen in the app GPS Status). But the circle has a diameter of 7m only in the highest zoom level. When zooming out the circle increases a little bit and has has constant size. It should become smaller and smaller when zooming out, shouldn't it?

**Reply by kds112** [reply]
<img src="https://github.com/devemux86/cruiser/assets/138766725/12224564-4d5b-4865-a21c-aed3fb5c08a8" width="50%"/>

**Reply by kds112** [reply]
Accuracy (gps error) of 5 km? The circle should be nearly invisible...

**Reply by kds112** [reply]
<img src="https://github.com/devemux86/cruiser/assets/138766725/129ff927-0ab5-47f3-b43a-7b62408e1aaf" width="50%"/>

And if zero satellites are locked I would expect a very large (infinite?) circle, but a "normal" size is shown.

**Reply by devemux86** [reply]
Thanks everyone for your tests!

> Any chance to change the order?

Do you mean the order of the panels at top-right?
This is something I already have in mind.

> But the circle has a diameter of 7m only in the highest zoom level. When zooming out the circle increases a little bit and has has constant size

The accuracy circle now appears from zoom 16. Before zoom 16, the fixed circle appears.
Indeed we should change it and always show the correct accuracy in all zooms.

> And if zero satellites are locked I would excpect a very large (infinite?) circle, but a "normal" size is shown.

What zoom is this?

The accuracy and the satellites are reported from the GPS sensor.
They are separate, what accuracy the sensor should report with zero satellites?

Because we should still show the last location.

**Reply by devemux86** [reply]
@kds112

Why the compass appears in this location and not at top-left?
Have you enabled any other setting in Cruiser or Android?

**Reply by kds112** [reply]
Yes, I meant the panel order.

Yes, please change the circle size at all zoom levels.

It is zoom level 16. Yes, well, it is a question how big the circle should be when no satellite is locked... Maybe a different colour?

**Reply by kds112** [reply]
I moved the compass because of the camera hole position when using full-screen mode.

**Reply by devemux86** [reply]
> Indeed we should change it and always show the correct accuracy in all zooms.

It will be available in the next Cruiser update.

**Reply by Ciutnik** [reply]
Yes, in b13 the accuracy circle looks better and carries more information.  I'll probably use it that way. Thanks! 👍

**Reply by kds112** [reply]
Yes indeed, it looks good! How do you show that it is the last known position and not the actual position with a locked gps signal? Different colour? Dashed line of the circle?

**Reply by devemux86** [reply]
Why do we need so much information about the GPS location?

If GPS works, it is the location from GPS, usually every 1 sec.

**Reply by kds112** [reply]
You are right, we don't need it, it was just an idea... :-)

**Reply by devemux86** [reply]
No problem, all ideas are useful to discuss! :slightly_smiling_face: 

We just need to analyze each one and see how it can be used or may complicate the app.
e.g. the location color is not predefined, users can change it in settings

**Reply by kds112** [reply]
Just a last question, I am sorry to bother you... Is the diameter of the accuracy circle equal to the gps error or twice the gps error? It should twice (the radius should be equal to the error). I guess that the circle is too small, but maybe I am wrong...

**Reply by devemux86** [reply]
@kds112 there is no problem!

The location circle radius is the estimated horizontal accuracy radius in meters, as reported by the GPS.
Usually it can be compared with the map scale bar to see if it is correct.

Have you seen something else?

**Reply by devemux86** [reply]
Note: location size can be changed in Cruiser settings.
But this does not affect the accuracy circle!

**Reply by Ciutnik** [reply]
@kds112 
Interesting observation! I did a simple test by running Cruiser and GPS Test at the same time. 

If such a test makes sense at all, it seems that the circle is drawn roughly well. Of course, we assume that while we can expect precision from GPS Test, we don't expect it from the accuracy circle 🙂

<img src="https://github.com/devemux86/cruiser/assets/122647782/b8154a9a-398b-4234-ab27-2f9c9e30c7ee" width="75%"/>

**Reply by devemux86** [reply]
Such GPS apps may read the NMEA values directly from the sensor.
Map / navigation apps usually work with the Android API (better battery).

Note that it is the horizontal accuracy (not vertical).
Also may be rounding cases here and there, conversion of distance to screen pixels, etc.

**Reply by Ciutnik** [reply]
@devemux86 
Sure! The test only confirmed the correctness of drawing a circle, knowing its advantages and disadvantages. I already liked him and he will remain my choice 👍

**Comment by devemux86** [comment, marked as answer]
GPS info panel and location accuracy are available in [Cruiser 3.0.32](https://play.google.com/store/apps/details?id=gr.talent.cruiser).
